src/stream/src/executor/rearranged_chain.rs

```diff
@@ -164,15 +169,14 @@ impl RearrangedChainExecutor {
 
             // Chain the `snapshot` and `upstream_rx` to get a unified `rearranged_chunks` stream.
             let rearranged_chunks = snapshot
-                .map(|result| result.map(RearrangedMessage::from))
+                .map(|result| result.map(RearrangedMessage::rearranged_from))
                 .chain(upstream_rx.map(Ok));
 
             // 5. Merge the rearranged barriers with chunks, with the priority of barrier.
-            let mut rearranged = select_with_strategy(
-                rearranged_barrier_rx.map(Ok),
-                rearranged_chunks,
-                |_state: &mut ()| stream::PollNext::Left,
-            );
+            let mut rearranged =
+                select_with_strategy(rearranged_barrier, rearranged_chunks, |_state: &mut ()| {
```

Comment:
Maybe we will need to employ techniques similar to https://github.com/singularity-data/risingwave/pull/2661 or wait for concurrent checkpoint, so that data in one epoch can be bound.

Reply:
Good catch! NTFS!